.measureText() returns TextMetrics object with rounded values for bounding box

rwlbuis opened the pull request:
#### b3809e07dc65e5678706c3ee334ca12930ebf129
<pre>
.measureText() returns TextMetrics object with rounded values for bounding box
<a href="https://bugs.webkit.org/show_bug.cgi?id=240213">https://bugs.webkit.org/show_bug.cgi?id=240213</a>

Reviewed by Said Abou-Hallawa.

Stop rounding intermediate calculations to integer values, potentially ending up with float values for TextMetrics bounding box.

Note that iOS/glib platforms seem to always round ascent/descent to integer values so have some different test results. Stopping the rounding
there would lead to a big amount of needed rebaselines, so it is left out of this PR.

* LayoutTests/fast/canvas/canvas-measureText-2-expected.txt:
* LayoutTests/fast/canvas/canvas-measureText-2.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::measureTextInternal):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleText const):
(WebCore::FontCascade::widthForComplexText const):

Canonical link: <a href="https://commits.webkit.org/288439@main">https://commits.webkit.org/288439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/782226a810ab418575d70edf47f34165624d3d45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88371 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34303 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2992 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10854 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64799 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22544 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86343 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2178 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75705 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45083 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29902 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33352 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89743 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10558 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73232 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71525 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72462 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17952 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16671 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15407 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10513 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15988 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10365 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12136 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1570,10 +1570,10 @@ float FontCascade::widthForSimpleText(const TextRun& run, SingleThreadWeakHashSe
-        glyphOverflow->left = ceilf(it.firstGlyphOverflow());
-        glyphOverflow->right = ceilf(it.lastGlyphOverflow());
+        glyphOverflow->top = std::max<double>(glyphOverflow->top, -it.minGlyphBoundingBoxY() - (glyphOverflow->computeBounds ? 0 : metricsOfPrimaryFont().intAscent()));
+        glyphOverflow->bottom = std::max<double>(glyphOverflow->bottom, it.maxGlyphBoundingBoxY() - (glyphOverflow->computeBounds ? 0 : metricsOfPrimaryFont().intDescent()));
```
Why are we still using `intAscent()` and `intDescent()` instead of of using `ascent()` and `descent()` since all the calculations are in `double`?

shallawa replied:
Also why this change is needed? The layout tests changes do not include changes in the text width. All of them are out the ascent, descent and the baseline.

rwlbuis replied:
> Why are we still using `intAscent()` and `intDescent()` instead of of using `ascent()` and `descent()` since all the calculations are in `double`?

Indeed we can just use `ascent()` and `descent()` here.

rwlbuis replied:
> Also why this change is needed? The layout tests changes do not include changes in the text width. All of them are out the ascent, descent and the baseline.

Do you mean there should be more tests for the horizontal values? I added some subtests for that in 2d.text.measure.actualBoundingBox.html.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.actualBoundingBox.html:
```diff
@@ -56,6 +56,9 @@ <h1>2d.text.measure.actualBoundingBox</h1>
 
+  _assert(Number.isInteger(ctx.measureText('D').actualBoundingBoxRight) == false, "Number.isInteger(ctx.measureText('D').actualBoundingBoxRight) == false");
+  _assert(Number.isInteger(ctx.measureText('ABCD').actualBoundingBoxRight) == false, "Number.isInteger(ctx.measureText('ABCD').actualBoundingBoxRight) == false");
+
```
I am sorry. I think I confused you by my question about `glyphOverflow` in `widthForSimpleText()` and `widthForComplexText()`. I just realized it is tested in the line above:
```
_assert(ctx.measureText('D').actualBoundingBoxRight >= 75, "ctx.measureText('D').actualBoundingBoxRight >= 75");
```
`CanvasRenderingContext2DBase::measureTextInternal()` uses `glyphOverflow` in calculating `actualBoundingBoxLeft` and `actualBoundingBoxRight`.
```
    metrics->setActualBoundingBoxLeft(glyphOverflow.left - offset.x());
    metrics->setActualBoundingBoxRight(fontWidth + glyphOverflow.right + offset.x());
```

shallawa replied:
So the new assertions you added are meaningless. The TextMetrics.idl says they are `double`. Please remove them.

shallawa commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1570,10 +1570,10 @@ float FontCascade::widthForSimpleText(const TextRun& run, SingleThreadWeakHashSe
-        glyphOverflow->bottom = std::max<int>(glyphOverflow->bottom, ceilf(it.maxGlyphBoundingBoxY()) - (glyphOverflow->computeBounds ? 0 : metricsOfPrimaryFont().intDescent()));
-        glyphOverflow->left = ceilf(it.firstGlyphOverflow());
-        glyphOverflow->right = ceilf(it.lastGlyphOverflow());
+        glyphOverflow->top = std::max<double>(glyphOverflow->top, -it.minGlyphBoundingBoxY() - (glyphOverflow->computeBounds ? 0 : metricsOfPrimaryFont().ascent()));
+        glyphOverflow->bottom = std::max<double>(glyphOverflow->bottom, it.maxGlyphBoundingBoxY() - (glyphOverflow->computeBounds ? 0 : metricsOfPrimaryFont().descent()));
+        glyphOverflow->left = it.firstGlyphOverflow();
+        glyphOverflow->right = it.lastGlyphOverflow();
```
Not new to this PR.

The name of this function `widthForSimpleText()` is confusing. It calculates the width of a TextRun. But in addition to this it can calculate an optional `GlyphOverflow` which can be used in calculating the boundingBox of a `TextMetrics`.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.actualBoundingBox.html:
```diff
@@ -46,5 +46,8 @@ <h1>2d.text.measure.actualBoundingBox</h1>
 
+  _assert(Number.isInteger(ctx.measureText('D').actualBoundingBoxRight) == false, "Number.isInteger(ctx.measureText('D').actualBoundingBoxRight) == false");
+  _assert(Number.isInteger(ctx.measureText('ABCD').actualBoundingBoxRight) == false, "Number.isInteger(ctx.measureText('ABCD').actualBoundingBoxRight) == false");
+
```
Please remove all the new added assertions.

rwlbuis committed:
.measureText() returns TextMetrics object with rounded values for bounding box
https://bugs.webkit.org/show_bug.cgi?id=240213

Reviewed by Said Abou-Hallawa.

Stop rounding intermediate calculations to integer values, potentially ending up with float values for TextMetrics bounding box.

Note that iOS/glib platforms seem to always round ascent/descent to integer values so have some different test results. Stopping the rounding
there would lead to a big amount of needed rebaselines, so it is left out of this PR.

* LayoutTests/fast/canvas/canvas-measureText-2-expected.txt:
* LayoutTests/fast/canvas/canvas-measureText-2.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines-expected.txt.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/text/2d.text.measure.baselines.worker-expected.txt.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::measureTextInternal):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleText const):
(WebCore::FontCascade::widthForComplexText const):

Canonical link: https://commits.webkit.org/288439@main
